src/gdb/gdb.c

```diff
@@ -238,6 +239,19 @@ while (1) {
 			break;
 		}
 		break;
+	/* Single step in the code */
+	case 's':
+		if (hex_to_int(&request, &addr))
+			sregs[DEBUG_PC] = addr;
+		/* leave debug just for one instruction */
+		sregs[ICOUNT] = 0xfffffffe;
+		sregs[ICOUNTLEVEL] = XCHAL_DEBUGLEVEL;
+		/* disable low level interrupts */
+		sregs[INTENABLE]  &= ~DISABLE_LOWER_INTERRUPTS_MASK;
+		write_sr(ICOUNTLEVEL);
+		write_sr(ICOUNT);
+		write_sr(INTENABLE);
```

Comment:
This single step logic is very xtensa specific, but the cmd parsing is generic. The single step code should be in arch/xtensa e.g. arch_gdb_single_step(generic args). This rule should apply to all architecture specific operations.